src/mainboard/amd/romecrb/src/main.rs

```diff
@@ -200,6 +203,110 @@ fn consdebug(w: &mut impl core::fmt::Write) -> () {
 }
 //global_asm!(include_str!("init.S"));
 
+// https://github.com/gz/rust-cpuid/issues/26#issuecomment-785280143
+fn amd_family_id(s: &FeatureInfo) -> u8 {
+    let base_family_id = s.family_id();
+    let extended_family_id = s.extended_family_id();
+    if base_family_id < 0xF {
+        base_family_id
+    } else {
+        base_family_id + extended_family_id
+    }
+}
+
+// https://github.com/gz/rust-cpuid/issues/26#issuecomment-785280143
+fn amd_model_id(s: &FeatureInfo) -> u8 {
+    let base_family_id = s.family_id();
+    let base_model_id = s.model_id();
+    let extended_model_id = s.extended_model_id();
+    if base_family_id < 0xF {
+        base_model_id
+    } else {
+        (extended_model_id << 4) | base_model_id
+    }
+}
+
+fn rome_ff_init(w: &mut impl core::fmt::Write) -> Result<(), &'static str> {
+    let hsmp = HSMP::new(0);
+    unsafe {
+        match hsmp.test(42) {
+            Ok(v) => {
+                write!(w, "HSMP test(42) result: {}\r\n", v).unwrap();
+            }
+            Err(e) => {
+                write!(w, "HSMP test(42) error: {}\r\n", e).unwrap();
+            }
+        }
+        match hsmp.smu_version() {
+            Ok((major, minor)) => {
+                write!(w, "HSMP smu version result: {}.{}\r\n", major, minor).unwrap();
+            }
+            Err(e) => {
+                write!(w, "HSMP smu version error: {}\r\n", e).unwrap();
+            }
+        }
+        match hsmp.interface_version() {
+            Ok((major, minor)) => {
+                write!(w, "HSMP interface version result: {}.{}\r\n", major, minor).unwrap();
+            }
+            Err(e) => {
+                write!(w, "HSMP interface version error: {}\r\n", e).unwrap();
+            }
+        }
+    }
+    Ok(())
+}
+
+fn amd_init(w: &mut impl core::fmt::Write) -> Result<(), &str> {
```

Comment:
and since we're using namespace scoping, and this is a cpu-level function, maybe
cpu_init

Reply:
Done